### DISCUSSION TITLE
config/view.php "cache" option does not work and causes views to render blank randomly

### DISCUSSION THREAD

**Discussion by stellar-scottreed** [laravel/framework#54031] [Q&A]
### Laravel Version

11.22

### PHP Version

8.3.14

### Description

I set the "cache" option in config/view.php to "false" (https://github.com/laravel/framework/pull/41859)
With this set this way, laravel is still generating views in storage/framework/views. Additionally, we had an extremely difficult to trace bug that this caused where views would randomly return an empty string. This was especially seen under concurrency.

### Steps To Reproduce

Look in the storage/framework/views folder when cache in config/view.php is set to false.

Also, I set up a nodejs script to fetch from a route with 5x requests at the same time and it would frequently return empty string instead of "it works!". When I removed the "cache" option in config/view.php this issue stopped occuring.

**Comment by rodrigopedra** [comment]
If you look at the linked PR's code (#41859), setting `cache` to `false` will make the blade compiler to always report the views as expired.

The compiler will always generate the PHP version of a view, that is what gets executed at the end, but when setting `cache` to `false` it will regenerate the file on every request.

From your description, it seems you are getting a race condition with many requests trying to write the same compiled view file, as each request thinks it is expired and tries to regenerate it again. 

As the view is not changed the hash for each request, its hash will result the same, and as each request race for the same file, some might be getting locked.

As stated in the linked PR, this option is meant to be used during development, where these high-concurrency scenarios are unlikely to happen.

What is the use-case where you need the view to be regenerated for every request?

Mind that, with the current implementation, skipping writing the generated PHP from a view to a file is not an option.

**Comment by rodrigopedra** [comment]
If all you want to avoid writing files to the disk, *in a scenario like deploying to an immutable container*, you can try setting the environment variable `VIEW_COMPILED_PATH` to a directory under `/tmp`, and make sure your `/tmp` is mounted as a `tmpfs` mount (assuming your server runs on Linux).

`tmpfs` means `/tmp` will be mounted to your RAM, so the blade compiler will "write" those compiled PHP files into your RAM, which is much faster than your disk and should improve the race condition scenario.

Mind that in a hugely concurrent scenario, you can still get locks. If there are many requests waiting to "write" the regenerated view into memory, the OS lock timeout might take longer than the requests awaiting.

I have no clue if you can have a partition or drive mapped to memory on Windows.

macOS should behave the same as Linux, but I am not sure.

---

EDIT: added a possible scenario where avoiding writing to disk might be wanted.

**Comment by stellar-scottreed** [comment]
@rodrigopedra Thank you for the quick reply! I really don't need "cache" set to false in production, and I've removed that. What happened is that I had a Laravel 5.4 codebase which had "cache" set to false (don't ask me why), which had no effect since the toggle wasn't introduced until Laravel 9. I upgraded to Laravel 11 and started having this issue in production, and finally traced it to that flag when I thought to disable view caching to try to fix it... and then I was very surprised to see view caching was disabled and that was the entire problem!

I wanted to submit a bug report in case there was something to prevent this for others, if not at least the search engines will index this and someone else can find this if they have this happen!

**Comment by ManicardiFrancesco** [comment]
Hi there,

I'm in a similar situation and was wondering if this feature is the right approach for my use case. In my application, I frequently use the Blade templating engine to send communications across various channels (WhatsApp, email, SMS, physical mail, etc.). This means I call:

`$compiledTemplate = Blade::render($myTemplate, $myParams);`
very frequently.
 However, since $myTemplate can vary with each message (they're often entirely custom), I end up with a very bloated ./storage/framework/views directory filled with cached views. I'm not using views for any other purpose since my Laravel instance is serving as a backend API only (never serving HTML from controllers).

I considered setting cache => false in config/view.php to prevent these files from being generated, but it looks like that option isn't intended for a high-concurrency production environment. Given that I also use workers to send multiple messages concurrently, what's the best approach to handle this scenario without overwhelming the storage with compiled views?

Thanks in advance for any guidance!

Best,
Francesco

**Reply by ManicardiFrancesco** [reply]
Answering my own question from 3 weeks ago with the solution I ended up going for:

just put in your schedule / crontab / whatever something that every once in a while runs php artisan view:clear

also - make sure to remember to restart your workers after you do this as if you don't they'll give you this error:

```
WARNING  require(storage/framework/views/3b0d71d8c2955fc983525bf81db2b0cc.php): Failed to open stream: No such file or directory in vendor/laravel/framework/src/Illuminate/Filesystem/Filesystem.php on line 123.


   Illuminate\View\ViewException  Failed opening required 'storage/framework/views/3b0d71d8c2955fc983525bf81db2b0cc.php' (include_path='.:/usr/share/php') (View: ).
```
